src/migration.py

```diff
@@ -0,0 +1,404 @@
+) -> types_.ActionReport:
+    """Write document content inside the docs directory.
+
+    Args:
+        file_meta: Information about migration file corresponding to a row in index table.
+        discourse: Client to the documentation server.
+        docs_path: The path to the docs directory to migrate all the documentation.
+
+    Raises:
+        MigrationError: if file_meta is of invalid metadata type.
+
+    Returns:
+        Migration report containing migration result.
+    """
+    match type(file_meta):
+        case types_.GitkeepMeta:
+            # To help mypy (same for the rest of the asserts), it is ok if the assert does not run
+            assert isinstance(file_meta, types_.GitkeepMeta)  # nosec
+            report = _migrate_gitkeep(gitkeep_meta=file_meta, docs_path=docs_path)
+        case types_.DocumentMeta:
+            assert isinstance(file_meta, types_.DocumentMeta)  # nosec
+            report = _migrate_document(
+                document_meta=file_meta, discourse=discourse, docs_path=docs_path
+            )
+        case types_.IndexDocumentMeta:
+            assert isinstance(file_meta, types_.IndexDocumentMeta)  # nosec
+            report = _migrate_index(index_meta=file_meta, docs_path=docs_path)
+        # Edge case that should not be possible.
+        case _:  # pragma: no cover
+            raise exceptions.MigrationError(
+                f"Internal error, no implementation for migration file, {file_meta=!r}"
+            )
+
+    logging.info("report: %s", report)
+    return report
+
+
+def _get_docs_metadata(
+    table_rows: typing.Iterable[types_.TableRow], index_content: str
+) -> itertools.chain[types_.MigrationFileMeta]:
+    """Get metadata for all documents to be migrated.
+
+    Args:
+        table_rows: Table rows from the index table.
+        index_content: Index content from index page.
+
+    Returns:
+        Metadata of files to be migrated.
+    """
+    index_doc = _index_file_from_content(content=index_content)
+    table_docs = _extract_docs_from_table_rows(table_rows=table_rows)
+    return itertools.chain((index_doc,), table_docs)
+
+
+def run(
+    table_rows: typing.Iterable[types_.TableRow],
+    index_content: str,
+    discourse: Discourse,
+    docs_path: Path,
+) -> None:
+    """Write table contents to the document directory.
+
+    Args:
+        table_rows: Iterable sequence of documentation structure to be migrated.
+        index_content: Main content describing the charm.
+        discourse: Client to the documentation server.
+        docs_path: The path to the docs directory containing all the documentation.
+
+    Raises:
+        MigrationError: if any migration report has failed.
+    """
+    valid_table_rows = (
+        valid_table_row for valid_table_row in _validate_table_rows(table_rows=table_rows)
+    )
+    migration_reports = (
+        _run_one(file_meta=document, discourse=discourse, docs_path=docs_path)
+        for document in _get_docs_metadata(
```

Comment:
A small refactor could be to make this a separate statement rather than nested within the generator, so maybe add a line `docs_metadata = get_docs_metadata...` and then this line becomes `for document in docs_metadata`. Do you think that would help with readability?

Reply:
Split it into few lines, thank you for the feedback!